### PR TITLE
Rework `Granularity` & `ModelWithSplitPoint`

### LINKED ISSUE #23: Make test for commons splitted model scaling

### LINKED ISSUE #48: `ModelWithSplitPoints.get_activations()` contain many padding activations.
## 🐛 Bug Report

When using `ModelWithSplitPoints.get_activations()` on a list of sentences, the sentences are padded to the maximum length. Hence, the activations are shaped (batch, max_seq_len, d_model), with activations corresponding to padding for some batch samples at the end of the seq_len dimension.

However, when the flattening strategy is used, the shape becomes (batch*max_seq_len, d_model), and padded activations become undiscernible from the others. They thus represent a high proportion of lines provided to concept_model fitting.

Furthermore, it becomes cumbersome to align these activations to the initial text for interpretation purposes.

### Code sample

```python
encoder_lm_splitter.get_activations(["short sentence", "surely a sentence with more tokens than the first one"]).shape
> torch.Size([2, 13, 312])
encoder_lm_splitter.get_activations(["short sentence", "surely a sentence with more tokens than the first one"], select_strategy="flatten").shape
> torch.Size([26, 312])
```

## Expected behavior

I don't know what kind of behavior we expect. However, I am sure the actual behavior is not desirable for the fit of the concept model.

I suggest that the shape becomes (seq1_len + ... seqn_len, model_d), and we manage the manipulation of these activations in the ModelWithSplitPoints.

### PR DESCRIPTION
**Pull request by AntoninPoche**
## Description

There are three main contributions in this pull request:
- Rework and expand `Granularity`
- Adapt `ModelWithSplitPoint` and `TopKInputs` to different granularities.
- Support batching and setting the device in `ModelWithSplitPoint`

Naturally, all the corresponding tests and documentation have been added. Furthermore, a test of `ModelWithSplitPoint` with many different models has been added.

Note that we depend on 'spacy' for the `Granularity.SENTENCE`. I tested it, and it is robust. It is not included in the dependencies by default, but can be installed via `uv pip install interpreto[spacy]`.

## Related Issue

fixes #23 by having test on many different models and batching. But we do not yet support parallelism.
fixes #48 by adding the `TOKEN` granularity.

## Type of Change

- 📚 docs / tutorials / dependencies update
- 🔧 Bug fix (non-breaking change which fixes an issue)
- 🥂 Improvement (non-breaking change which improves an existing feature)
- 🚀 New feature (non-breaking change which adds functionality)

## Checklist

- [x] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/FOR-sight-ai/interpreto/blob/main/CODE_OF_CONDUCT.md) document.
- [x] I've read the [`CONTRIBUTING.md`](https://github.com/FOR-sight-ai/interpreto/blob/main/docs/contributing.md) guide.
- [x] I've successfully run the style checks using `make lint`.
- [x] I've written tests for all new methods and classes that I created and successfully ran `make test`.
- [x] I've written the docstring in Google format for all the methods and classes that I used.

### CODE REVIEW

File: docs/api/concepts/interpretations/overview.md [hunk 1]
```diff
@@ -1,7 +0,0 @@
----
-icon: material/text-short
----
-
-# Concept Interpretations
-
-TODO: Add a brief description of the concept interpretations.
```
Comment (fanny-jourdan) [inline docs/api/concepts/interpretations/overview.md:7]: Maybe you forgot to do the TODO and wanted to do it before the PR?

File: docs/api/concepts/overview.md [hunk 2]
```diff
@@ -1,39 +1,88 @@
...
+### Step 2: Compute the model activations on the split_point
 
-integrated_gradient_kwargs = {"step": 10, "baseline": "[MASK]"}
-concept_input_attribution = concept_explainer.input_concept_attribution(attribution_method=IntegratedGradient, examples, **integrated_gradient_kwargs)
+At this step, there are two important parameters:
 
-TopKtokens_kwargs = {todo}
-concept_interpretation = concept_explainer.interpret(interpretation_method=TopKtokens, examples, **TopKtokens_kwargs)
+- `select_strategy`: specifies which of the `(n, l, d)` activations to return.
+It can be one of `ALL`, `CLS`, `ALL_TOKENS`, `TOKEN`, `WORD`, `SENTENCE`, `SAMPLE`.
```
Comment (fanny-jourdan) [inline docs/api/concepts/overview.md]: maybe put the granularities already present and the others in “soon” or something like that?

File: docs/api/concepts/overview.md [hunk 3]
```diff
@@ -1,39 +1,88 @@
...
+Use `select_strategy=ModelWithSplitPoints.activation_strategies.TOKEN` to specify it.
 
-```
+- `aggregation_strategy`: the mode used for inference.
+It can be one of `SUM`, `MEAN`, `MAX`, or `SIGNED_MAX`.
```
Comment (fanny-jourdan) [inline docs/api/concepts/overview.md:55]: I'm not sure what SIGNED_MAX means, do you specify at some point? (Maybe not there, but at least somewhere).

Comment (AntoninPoche) [reply docs/api/concepts/overview.md:55]: I will describe it in the documentation.

File: docs/api/concepts/overview.md (hunk 2 above)
Comment (AntoninPoche) [reply docs/api/concepts/overview.md]: Not sure what you mean.

File: docs/api/concepts/interpretations/overview.md (hunk 1 above)
Comment (AntoninPoche) [reply docs/api/concepts/interpretations/overview.md:7]: For now, I remove it because there is only one method. Thus, it does not make sense to have an overview.

But I will clarify the documentation of `TopKInputs`.

File: docs/api/concepts/overview.md (hunk 2 above)
Comment (fanny-jourdan) [reply docs/api/concepts/overview.md]: I thought you only had: `ALL`, `ALL_TOKENS`, `TOKEN`, `WORD`, `SENTENCE`, `SAMPLE`.

and not yet `CLS` ?

Comment (AntoninPoche) [reply docs/api/concepts/overview.md]: `CLS` is indeed available for `ModelWithSplitPoint.get_activations()`, in fact, it just corresponds to taking the activations of the first token.

(n, l, d)[:, 0, :] -> (n, d)

Comment (fanny-jourdan) [reply docs/api/concepts/overview.md]: It's not clear because it looks like CLS = clauses. Maybe rename it Token_CLS ?

Comment (AntoninPoche) [reply docs/api/concepts/overview.md]: Indeed, now that you say it, `CLS` can be seen as clause, I will clarify it!

File: interpreto/model_wrapping/model_with_split_points.py [hunk 4]
```diff
@@ -99,11 +175,13 @@ class ModelWithSplitPoints(LanguageModel):
...
-        split_points: str | int | list[str | int] | tuple[str | int],
+        split_points: str | int | list[str] | list[int] | tuple[str] | tuple[int],
         *args: tuple[Any],
         model_autoclass: str | type[AutoModel] | None = None,
         tokenizer: PreTrainedTokenizer | PreTrainedTokenizerFast | None = None,
         config: PretrainedConfig | None = None,
+        batch_size: int = 1,
+        device_map: str | None = None,
         **kwargs,
     ) -> None:
```
Comment (fanny-jourdan) [inline interpreto/model_wrapping/model_with_split_points.py:186]: when initializing, check that if you've given anything other than “str” to the model, then the tokenizer is not set to None.

File: interpreto/model_wrapping/model_with_split_points.py [hunk 5]
```diff
@@ -212,34 +302,190 @@ def _generate(
...
+                In the case of a `torch.Tensor`, we assume a batch dimension and token ids.
+            select_strategy (ActivationSelectionStrategy): Selection strategy for activations.
 
                 Options are:
+                - ``ModelWithSplitPoints.activation_strategies.ALL``:
+                    the activations are returned as is ``(batch, seq_len, d_model)``.
+                    They are padded manually so that each batch of activations can be concatenated.
+
+                - ``ModelWithSplitPoints.activation_strategies.CLS``:
+                    only the first token (e.g. ``[CLS]``) activation is returned ``(batch, d_model)``.
```
Comment (fanny-jourdan) [inline interpreto/model_wrapping/model_with_split_points.py]: We need a check to ensure that if we request the CLS token, it's because we have a model where the name ends with: ForSequenceClassification.

File: interpreto/model_wrapping/model_with_split_points.py [hunk 6]
```diff
@@ -212,34 +302,190 @@ def _generate(
...
+                Applied for `WORD`, `SENTENCE` and `SAMPLE` activation strategies.
+                Token activations of shape  n * (l, d) are aggregated on the sequence length dimension.
+                The concatenated into (ng, d) tensors.
+                Existing strategies are:
+
+                - ``ModelWithSplitPoints.aggregation_strategies.SUM``:
+
+                - ``ModelWithSplitPoints.aggregation_strategies.MEAN``:
+
+                - ``ModelWithSplitPoints.aggregation_strategies.MAX``:
```
Comment (fanny-jourdan) [inline interpreto/model_wrapping/model_with_split_points.py]: even if the meaning is obvious, maybe put a little sentence to say what everyone does?

File: tests/attributions/test_base_generation.py [hunk 7]
```diff
@@ -38,11 +38,11 @@ def create_targets_test(tokenizer):
...
 
     # TensorMapping:
-    target2a = tokenizer(target1a, return_tensors="pt")
-    target2b = tokenizer([target1b], return_tensors="pt")
+    target2a = tokenizer(target1a, return_tensors="pt", return_offsets_mapping=True)
+    target2b = tokenizer([target1b], return_tensors="pt", return_offsets_mapping=True)
 
     # TensorMapping with multiple elements:
-    target2c = tokenizer([target1a, target1b], return_tensors="pt", padding=True)
+    target2c = tokenizer([target1a, target1b], return_tensors="pt", padding=True, return_offsets_mapping=True)
```
Comment (fanny-jourdan) [inline tests/attributions/test_base_generation.py:45]: in fact, it might be interesting to keep an example where offsetmapping is not True, so that we can see if our code manages to reintegrate it if the user didn't give it in the first place.

File: interpreto/commons/granularity.py [hunk 8]
```diff
@@ -57,158 +68,264 @@ class Granularity(Enum):
...
+        positions of the tokens that compose one granularity unit:
 
-    @staticmethod
-    def get_length(tokens_ids: TensorMapping, granularity: Granularity | None = None) -> torch.Tensor:
-        """
-        Returns the length of the sequences according to the granularity level
+        * **ALL_TOKENS**  ``[[0], [1], ..., [Lp-1]]``
+        * **TOKEN**       same as above but *without* special-token positions
+        * **WORD**        one sub-list per word, e.g.
+                           ``[[1], [2, 3], [4]]`` for “a| beau|tiful| day”
```
Comment (fanny-jourdan) [inline interpreto/commons/granularity.py]: can't you get SENTENCE here?

File: interpreto/commons/granularity.py [hunk 9]
```diff
@@ -57,158 +68,264 @@ class Granularity(Enum):
...
+
+                if not _HAS_SPACY:
+                    raise ModuleNotFoundError(
+                        "spaCy is needed for sentence granularity.  Install with: `uv pip install spacy`"
                     )
-                    + 1
-                )
+
+                n_inputs = inputs["input_ids"].shape[0]  # type: ignore
+                offset_maps = inputs["offset_mapping"]  # (n, lp, 2)
```
Comment (fanny-jourdan) [inline interpreto/commons/granularity.py]: I dont understand this

File: interpreto/attributions/perturbations/base.py [hunk 10]
```diff
@@ -133,15 +134,12 @@ def perturb(self, inputs: TensorMapping) -> tuple[TensorMapping, torch.Tensor |
         """
         Method called when we ask the perturbator to perturb a mapping of tensors, generally the output of a tokenizer
         The mapping should be similar to mappings returned by the tokenizer.
-        It should at least have "input_ids", "attention_mask" and "offset_mappings" keys
+        It should at least have "input_ids" and "attention_mask"
```
Comment (fanny-jourdan) [inline interpreto/attributions/perturbations/base.py]: since you're removing offset_mappings here, do we need it anywhere else, or can we remove it from the tokenizer when we do it in explain?

File: tests/concepts/interpretation/test_topk_inputs.py [hunk 11]
```diff
@@ -73,8 +73,8 @@ def test_topk_inputs_from_activations(splitted_encoder_ml: ModelWithSplitPoints)
         model_with_split_points=splitted_encoder_ml,
         split_point=split,
         concept_model=concept_model,
-        granularity=Granularities.TOKENS,
-        source=InterpretationSources.CONCEPTS_ACTIVATIONS,
+        granularity=TopKInputs.granularities.TOKEN,
```
Comment (fanny-jourdan) [inline tests/concepts/interpretation/test_topk_inputs.py]: given that granularity is something we need everywhere, wouldn't it be more natural to directly call Granularity.TOKENS with an import like:
from interpreto import Granularity

File: tests/attributions/test_base_generation.py (hunk 7 above)
Comment (AntoninPoche) [reply tests/attributions/test_base_generation.py:45]: It is quite complex; it would require reassembling the input and then tokenizing it again.

In fact, I would argue that supporting too many input formats might be detrimental. If we are forced to manipulate a lot of the inputs, we might create bugs we did not see coming...

File: interpreto/commons/granularity.py (hunk 8 above)
Comment (AntoninPoche) [reply interpreto/commons/granularity.py]: Sure, I just forgot it ^^

File: interpreto/commons/granularity.py (hunk 9 above)
Comment (AntoninPoche) [reply interpreto/commons/granularity.py]: In this case, the user specified the SENTENCE granularity; several things are required: a fast tokenizer, offset mappings in the inputs, and Spacy installed.

If there are more questions, the best would be a call.

File: interpreto/attributions/perturbations/base.py (hunk 10 above)
Comment (AntoninPoche) [reply interpreto/attributions/perturbations/base.py]: It was a mistake to delete it from here. I removed it before adding the SENTENCE granularity and forgot to put it back.

The `offsets_mapping` are necessary only for the `SENTENCE` granularity.

I chose to add it by default when tokenizing in our code but do not enforce it when the user passes `BatchEncoding`. It will only raise an error saying to add the offsets mapping if the user passes `BatchEncoding`, selects `SENTENCE` granularity, and forgets the offsets mapping.

Should we make a case to only add offsets_mapping in our code if the granularity is `SENTENCE`? I think it would make the code harder to read, and the impact on the memory is really thin.

File: tests/concepts/interpretation/test_topk_inputs.py (hunk 11 above)
Comment (AntoninPoche) [reply tests/concepts/interpretation/test_topk_inputs.py]: Here, there are additional possible values. The possible granularities here correspond to the `ActivationSelectionStrategies`. Should I change the name. We can also change the name in `ModelWithSplitPoint`.
